File: MIDTERM-LAB-1/CustomerSegmentation.java

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;

public class CustomerSegmentation {

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        File file;

        // ── 1. FILE PATH INPUT & VALIDATION LOOP ──────────────────────────────
        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = input.nextLine().trim();
            file = new File(path);

            if (!file.exists() || !file.isFile()) {
                System.out.println("Invalid file path. File does not exist. Please try again.\n");
            } else if (!file.canRead()) {
                System.out.println("File exists but cannot be read. Check permissions.\n");
            } else if (!path.toLowerCase().endsWith(".csv")) {
                System.out.println("File does not appear to be a CSV. Please provide a .csv file.\n");
            } else {
                System.out.println("File found. Processing...\n");
                break;
            }
        }

        // ── 2. LOAD & AGGREGATE DATA ──────────────────────────────────────────
        Map<String, Double> publisherSales = new HashMap<>();
        int linesRead = 0;
        int parseErrors = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line;
            boolean isHeader = true;

            while ((line = br.readLine()) != null) {
                if (isHeader) { isHeader = false; continue; } // skip header

                String[] cols = line.split(",", -1);
                if (cols.length < 8) { parseErrors++; continue; }

                try {
                    String publisher = cols[4].trim();
                    String salesStr  = cols[7].trim();

                    if (publisher.isEmpty() || salesStr.isEmpty()) { parseErrors++; continue; }

                    double sales = Double.parseDouble(salesStr);
                    publisherSales.merge(publisher, sales, Double::sum);
                    linesRead++;

                } catch (NumberFormatException e) {
                    parseErrors++;
                }
            }

        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            input.close();
            return;
        }

        // ── 3. BUILD DataRecord LIST & SEGMENT ────────────────────────────────
        Map<String, List<DataRecord>> segments = new TreeMap<>();
        segments.put("Platinum", new ArrayList<>());
        segments.put("Gold",     new ArrayList<>());
        segments.put("Silver",   new ArrayList<>());
        segments.put("Bronze",   new ArrayList<>());

        for (Map.Entry<String, Double> entry : publisherSales.entrySet()) {
            DataRecord record = new DataRecord(entry.getKey(), entry.getValue());
            segments.get(record.getSegment()).add(record);
        }

        // ── 4. DISPLAY RESULTS ────────────────────────────────────────────────
        String border = "=".repeat(100);
        String divider = "-".repeat(100);

        System.out.println(border);
        System.out.println("  CUSTOMER SEGMENTATION REPORT — Video Game Sales 2024");
        System.out.println("  Dataset: vgchartz-2024.csv  |  Records Processed: " + linesRead
                + "  |  Parse Errors: " + parseErrors);
        System.out.println(border);
        System.out.printf("  %-12s | %-8s | %s%n", "SEGMENT", "COUNT", "THRESHOLD (purchases)");
        System.out.println(divider);
        System.out.printf("  %-12s | %-8d | > 100,000%n",   "Platinum", segments.get("Platinum").size());
        System.out.printf("  %-12s | %-8d | 50,000 – 100,000%n", "Gold", segments.get("Gold").size());
        System.out.printf("  %-12s | %-8d | 10,000 – 49,999%n",  "Silver", segments.get("Silver").size());
        System.out.printf("  %-12s | %-8d | < 10,000%n",    "Bronze", segments.get("Bronze").size());
        System.out.println(border);

        String[] order = {"Platinum", "Gold", "Silver", "Bronze"};
        for (String seg : order) {
            List<DataRecord> group = segments.get(seg);
            System.out.printf("%n  ┌─ %s SEGMENT (%d publishers)%n", seg.toUpperCase(), group.size());
            System.out.println("  " + divider);
            if (group.isEmpty()) {
                System.out.println("  │  (no publishers in this segment)");
            } else {
                group.sort((a, b) -> Double.compare(b.getTotalSales(), a.getTotalSales()));
                for (DataRecord r : group) {
                    System.out.println("  │  " + r);
                }
            }
            System.out.println("  └" + "─".repeat(99));
        }

        System.out.println("\n  END OF REPORT");
        System.out.println(border);
        input.close();
    }
}

File: MIDTERM-LAB-1/customerSegmentation.js
'use strict';

const fs       = require('fs');
const path     = require('path');
const readline = require('readline');

// ── Readline interface ────────────────────────────────────────────────────────
const rl = readline.createInterface({
    input:  process.stdin,
    output: process.stdout,
});

// ── DataRecord class (mirrors Java counterpart) ───────────────────────────────
class DataRecord {
    constructor(publisher, totalSales) {
        this.publisher  = publisher;
        this.totalSales = totalSales;           // in millions
        this.purchases  = totalSales * 1000;    // convert → thousands
    }

    getSegment() {
        if (this.purchases > 100000)      return 'Platinum';
        if (this.purchases >= 50000)      return 'Gold';
        if (this.purchases >= 10000)      return 'Silver';
        return 'Bronze';
    }

    toString() {
        return (
            this.publisher.padEnd(40) +
            ` | Sales: ${this.totalSales.toFixed(2).padStart(10)} M` +
            ` | Purchases: ${this.purchases.toFixed(0).padStart(12)}` +
            ` | Segment: ${this.getSegment()}`
        );
    }
}

// ── Utility: parse a CSV line respecting quoted fields ───────────────────────
function parseCsvLine(line) {
    const result = [];
    let current  = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (ch === '"') {
            inQuotes = !inQuotes;
        } else if (ch === ',' && !inQuotes) {
            result.push(current.trim());
            current = '';
        } else {
            current += ch;
        }
    }
    result.push(current.trim());
    return result;
}

// ── Core processing function ─────────────────────────────────────────────────
function processFile(filePath) {
    let rawData;

    try {
        rawData = fs.readFileSync(filePath, 'utf8');
    } catch (err) {
        console.error(`Error reading file: ${err.message}`);
        rl.close();
        return;
    }

    const lines      = rawData.split(/\r?\n/);
    const publisherSales = {};
    let linesRead    = 0;
    let parseErrors  = 0;

    // Skip header (index 0)
    for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        try {
            const cols      = parseCsvLine(line);
            if (cols.length < 8) { parseErrors++; continue; }

            const publisher = cols[4];
            const salesRaw  = cols[7];

            if (!publisher || !salesRaw) { parseErrors++; continue; }

            const sales = parseFloat(salesRaw);
            if (isNaN(sales))            { parseErrors++; continue; }

            publisherSales[publisher] = (publisherSales[publisher] || 0) + sales;
            linesRead++;
        } catch {
            parseErrors++;
        }
    }

    // Build DataRecord objects and segment
    const segments = { Platinum: [], Gold: [], Silver: [], Bronze: [] };

    for (const [publisher, totalSales] of Object.entries(publisherSales)) {
        const record = new DataRecord(publisher, totalSales);
        segments[record.getSegment()].push(record);
    }

    // Sort each segment by sales descending
    for (const seg of Object.values(segments)) {
        seg.sort((a, b) => b.totalSales - a.totalSales);
    }

    // ── Display results ───────────────────────────────────────────────────────
    const border  = '='.repeat(100);
    const divider = '-'.repeat(100);

    console.log('\n' + border);
    console.log('  CUSTOMER SEGMENTATION REPORT — Video Game Sales 2024');
    console.log(`  Dataset: ${path.basename(filePath)}  |  Records Processed: ${linesRead}  |  Parse Errors: ${parseErrors}`);
    console.log(border);
    console.log(`  ${'SEGMENT'.padEnd(12)} | ${'COUNT'.padEnd(8)} | THRESHOLD (purchases)`);
    console.log(divider);
    console.log(`  ${'Platinum'.padEnd(12)} | ${String(segments.Platinum.length).padEnd(8)} | > 100,000`);
    console.log(`  ${'Gold'.padEnd(12)} | ${String(segments.Gold.length).padEnd(8)} | 50,000 – 100,000`);
    console.log(`  ${'Silver'.padEnd(12)} | ${String(segments.Silver.length).padEnd(8)} | 10,000 – 49,999`);
    console.log(`  ${'Bronze'.padEnd(12)} | ${String(segments.Bronze.length).padEnd(8)} | < 10,000`);
    console.log(border);

    for (const seg of ['Platinum', 'Gold', 'Silver', 'Bronze']) {
        const group = segments[seg];
        console.log(`\n  ┌─ ${seg.toUpperCase()} SEGMENT (${group.length} publishers)`);
        console.log('  ' + divider);
        if (group.length === 0) {
            console.log('  │  (no publishers in this segment)');
        } else {
            group.forEach(r => console.log('  │  ' + r.toString()));
        }
        console.log('  └' + '─'.repeat(99));
    }

    console.log('\n  END OF REPORT');
    console.log(border + '\n');

    rl.close();
}

// ── File path input & validation loop ────────────────────────────────────────
function askFilePath() {
    rl.question('Enter dataset file path: ', function (inputPath) {
        const trimmed = inputPath.trim();

        if (!fs.existsSync(trimmed)) {
            console.log('Invalid file path. File does not exist. Please try again.\n');
            askFilePath();
            return;
        }

        const stat = fs.statSync(trimmed);
        if (!stat.isFile()) {
            console.log('Path is not a file. Please provide a CSV file path.\n');
            askFilePath();
            return;
        }

        if (!trimmed.toLowerCase().endsWith('.csv')) {
            console.log('File does not appear to be a CSV. Please provide a .csv file.\n');
            askFilePath();
            return;
        }

        console.log('File found. Processing...\n');
        processFile(trimmed);
    });
}

// ── Entry point ───────────────────────────────────────────────────────────────
console.log('='.repeat(100));
console.log('  PROGRAMMING 2 – Customer Segmentation Machine Problem');
console.log('  University of Perpetual Help System DALTA – Molino Campus');
console.log('='.repeat(100));
askFilePath();
